src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php

```diff
@@ -47,29 +45,25 @@ public function applyToCollection(
             return;
         }
 
-        if (isset($context['filters'])){
-
+        if (isset($context['filters'])) {
             $taxonCode = $context['filters']['productTaxons.taxon.code'] ?? null;
             $this->addSortingToQuery($queryBuilder, $taxonCode);
         }
     }
 
     private function addSortingToQuery(QueryBuilder $queryBuilder, ?string $taxonCode): void
     {
-        if (!$taxonCode){
+        if (!$taxonCode) {
             return;
         }
 
-        $taxon = $this->taxonRepository->findOneBy(['code' => $taxonCode]);
-
-        if (!$taxon) {
-            return;
-        }
+        $rootAlias = $queryBuilder->getRootAliases()[0];
 
         $queryBuilder
-            ->leftJoin('o.productTaxons', 'productTaxons', 'WITH', 'productTaxons.taxon = :taxonId')
-            ->orderBy('productTaxons.position', 'ASC')
-            ->setParameter('taxonId', $taxon->getId())
+            ->leftJoin(sprintf('%s.productTaxons', $rootAlias), 'pt', 'WITH', sprintf('pt.product = %s.id', $rootAlias))
```

Comment:
`pt` and `tax` should also be properly aliased (`$queryNameGenerator->generateParameterName`)